463 support os root

sean9999 opened the pull request:
add support for os.Root.
see: https://go.dev/blog/osroot.

This code mainly sticks to just the task at hand, however in order to make CI tests pass, I had to do some very minor clean-ups.

I introduce the interface `Root`, which satisfies the same method-set as os.Root.

Root is basically a "chroot". It offers a root-relative view into a filesystem that you can't escape.

Root is powered by *rootFs, which simply wraps *BasePathFs, changing the signature slightly so that it satisfies the interface, and the functionality that os.Root provides.

sean9999 committed:
add os.Root functionality.

- add method to Fs interface
- implement the method in all implementations
- added OpenRoot and OpenInRoot
- tests
- upgrade go version and some transient deps with vulns

sean9999 committed:
add os.Root functionality.

- add method to Fs interface
- implement the method in all implementations
- added OpenRoot and OpenInRoot
- tests
- upgrade go version and some transient deps with vulns

sean9999 committed:
add os.Root functionality.

- add method to Fs interface
- implement the method in all implementations
- added OpenRoot and OpenInRoot
- tests
- upgrade go version and some transient deps with vulns

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/spf13/afero?pullRequest=507) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/spf13/afero?pullRequest=507) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: sean9999<br/>:x: Sean Macdonald<br/><hr/>**Sean Macdonald** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/spf13/afero?pullRequest=507) it.</sub>

sean9999 committed:
fix linting problems

sean9999 committed:
fix linting problems

sean9999 committed:
fix linting problems

MarkRosemaker committed:
fix spelling

sean9999 commented:
> Thanks @sean9999
> 
> I think this is a great addition!
> 
> If you don't mind, I'd like to wait for 1.25 to be released: we usually try to support two versions.

It's time, boss! =)
https://go.dev/blog/go1.25

sean9999 committed:
make CI test use versions >= v1.24
use toolchain directive

sean9999 committed:
make CI test use versions >= v1.24
use toolchain directive

sean9999 committed:
make CI test use versions >= v1.24
use toolchain directive

dependabot[bot] committed:
build(deps): bump actions/dependency-review-action from 4.7.1 to 4.7.3

Bumps [actions/dependency-review-action](https://github.com/actions/dependency-review-action) from 4.7.1 to 4.7.3.
- [Release notes](https://github.com/actions/dependency-review-action/releases)
- [Commits](https://github.com/actions/dependency-review-action/compare/da24556b548a50705dd671f47852072ea4c105d9...595b5aeba73380359d98a5e087f648dbb0edce1b)

---
updated-dependencies:
- dependency-name: actions/dependency-review-action
  dependency-version: 4.7.3
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump actions/setup-go from 5.5.0 to 6.0.0

Bumps [actions/setup-go](https://github.com/actions/setup-go) from 5.5.0 to 6.0.0.
- [Release notes](https://github.com/actions/setup-go/releases)
- [Commits](https://github.com/actions/setup-go/compare/d35c59abb061a4a6fb18e82ac0862c26744d6ab5...44694675825211faa026b3c33043df3e48a5fa00)

---
updated-dependencies:
- dependency-name: actions/setup-go
  dependency-version: 6.0.0
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump github/codeql-action from 3.29.7 to 3.30.1

Bumps [github/codeql-action](https://github.com/github/codeql-action) from 3.29.7 to 3.30.1.
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/51f77329afa6477de8c49fc9c7046c15b9a4e79d...f1f6e5f6af878fb37288ce1c627459e94dbf7d01)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-version: 3.30.1
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump actions/checkout from 4.2.2 to 5.0.0

Bumps [actions/checkout](https://github.com/actions/checkout) from 4.2.2 to 5.0.0.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/11bd71901bbe5b1630ceea73d27597364c9af683...08c6903cd8c0fde910a37f88322edcfb5dd907a8)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: 5.0.0
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

sagikazarmark committed:
ci: update ci

Signed-off-by: Mark Sagi-Kazar <mark.sagikazar@gmail.com>

sagikazarmark committed:
chore: update deps

Signed-off-by: Mark Sagi-Kazar <mark.sagikazar@gmail.com>

sagikazarmark committed:
Update afero to v1.15.0

Signed-off-by: Mark Sagi-Kazar <mark.sagikazar@gmail.com>

dependabot[bot] committed:
build(deps): bump github/codeql-action from 3.30.1 to 3.30.5

Bumps [github/codeql-action](https://github.com/github/codeql-action) from 3.30.1 to 3.30.5.
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/f1f6e5f6af878fb37288ce1c627459e94dbf7d01...3599b3baa15b485a2e49ef411a7a4bb2452e7f93)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-version: 3.30.5
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump actions/dependency-review-action from 4.7.3 to 4.8.0

Bumps [actions/dependency-review-action](https://github.com/actions/dependency-review-action) from 4.7.3 to 4.8.0.
- [Release notes](https://github.com/actions/dependency-review-action/releases)
- [Commits](https://github.com/actions/dependency-review-action/compare/595b5aeba73380359d98a5e087f648dbb0edce1b...56339e523c0409420f6c2c9a2f4292bbb3c07dd3)

---
updated-dependencies:
- dependency-name: actions/dependency-review-action
  dependency-version: 4.8.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump ossf/scorecard-action from 2.4.2 to 2.4.3

Bumps [ossf/scorecard-action](https://github.com/ossf/scorecard-action) from 2.4.2 to 2.4.3.
- [Release notes](https://github.com/ossf/scorecard-action/releases)
- [Changelog](https://github.com/ossf/scorecard-action/blob/main/RELEASE.md)
- [Commits](https://github.com/ossf/scorecard-action/compare/05b42c624433fc40578a4040d5cf5e36ddca8cde...4eaacf0543bb3f2c246792bd56e8cdeffafb205a)

---
updated-dependencies:
- dependency-name: ossf/scorecard-action
  dependency-version: 2.4.3
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump github/codeql-action from 3.30.5 to 4.30.8

Bumps [github/codeql-action](https://github.com/github/codeql-action) from 3.30.5 to 4.30.8.
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/3599b3baa15b485a2e49ef411a7a4bb2452e7f93...f443b600d91635bebf5b0d9ebc620189c0d6fba5)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-version: 4.30.8
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

sean9999 committed:
update deps

sean9999 committed:
update deps

spf13 commented:
The concept is good — `os.Root` (Go 1.24) fits naturally with afero's layered FS model. But adding `OpenRoot` to the `Fs` interface is a breaking change: every external `afero.Fs` implementation breaks at compile time. That's the same problem as adding a method to `File`, and the fix is the same pattern:

```go
type RootOpener interface {
    OpenRoot(name string) (Root, error)
}
```

Callers type-assert. The built-in backends implement it where it makes sense.

Other blocking issues:
1. `go.mod` minimum bumped to Go 1.24 — this constrains all afero users. If `os.Root` is behind a `RootOpener` interface, the minimum version bump can stay local to callers who use it (or be gated behind a build tag).
2. testify added to the core `go.mod` — keep test-only dependencies in a separate `go.mod` or use stdlib `testing`/`testing/quick`.
3. CI must pass on current master before we can merge.

Please re-submit with the `RootOpener` interface approach, with CI green.
